library/core/src/iter/adapters/map_windows.rs

```diff
@@ -0,0 +1,52 @@
+use crate::fmt;
+
+/// An iterator over the mapped windows of another iterator.
+///
+/// This `struct` is created by the [`Iterator::map_windows`]. See its
+/// documentation for more information.
+#[must_use = "iterators are lazy and do nothing unless consumed"]
+#[unstable(feature = "iter_map_windows", reason = "recently added", issue = "none")]
+pub struct MapWindows<I: Iterator, F, const N: usize> {
+    iter: I,
+    f: F,
+    buffer: Option<[I::Item; N]>,
+}
+
+impl<I: Iterator, F, const N: usize> MapWindows<I, F, N> {
+    pub(in crate::iter) fn new(mut iter: I, f: F) -> Self {
+        assert!(N > 0, "array in `Iterator::map_windows` must contain more than 0 elements");
+
+        let buffer = crate::array::collect_into_array(&mut iter);
+        Self { iter, f, buffer }
+    }
+}
+
+#[unstable(feature = "iter_map_windows", reason = "recently added", issue = "none")]
+impl<I, F, R, const N: usize> Iterator for MapWindows<I, F, N>
+where
+    I: Iterator,
+    F: FnMut(&[I::Item; N]) -> R,
+{
+    type Item = R;
+    fn next(&mut self) -> Option<Self::Item> {
+        let buffer = self.buffer.as_mut()?;
+        let out = (self.f)(buffer);
+
+        // Advance iterator
+        if let Some(next) = self.iter.next() {
+            buffer.rotate_left(1);
+            buffer[N - 1] = next;
```

Comment:
This is slightly wasteful in theory: the first element does not need to be copied to the last slot as it's overwritten later anyway. I *expect* this to be optimized out anyway, but I haven't checked.

I haven't found a better non-`unsafe` way to do this. `copy_within` requires `T: Copy`, `swap` would work, but I doubt doing a bunch of `swaps` is faster than the current version. Of course, the `unsafe` would be straight-forward but I still wanted to avoid it for now. Any ideas or opinions?

Reply:
Using unsafe in std lib is acceptable if it leads to a performance advantage.

Reply:
Is it possible to just write to the first element then rotate?

Reply:
Writing to the first element doesn't seem to change that much compared to the current approach.

I played a bit on godbolt and I found 3 situations:
* They're optimized almost the same way. This often happen for [tiny `T`s](https://godbolt.org/z/3d7K6P), but sometimes also for [bigger ones](https://godbolt.org/z/5onM31). 
* The `unsafe` version may be optimized to 4-5x less assembly. This often happen for [bigger `T`s but small `N`](https://godbolt.org/z/3afGs7).
* The `unsafe` version is optimized to a `memmove` call and the safe version to either a lot of moves or a loop. This often happen for [bigger `T`s and bigger `N`](https://godbolt.org/z/Tx7xv3)

Reply:
I think that, if `.rotate_right(1)` and `.rotate_left(1)` don't optimize to the obvious `memmove`, then that's a bug in those methods that should be fixed.

(I vaguely recall someone on Discord was going to file me a bug about this...)

Reply:
@SkiFire13 Thanks for testing this stuff. But I agree with @scottmcm in that this should probably optimize to the optimal code basically. The implementation of this can always be changed later, so I wouldn't say it is hugely important to get this as efficient as possible in this PR. Especially before we even agreed we want this method.